src/crimson/net/SocketConnection.cc

```diff
@@ -69,6 +66,28 @@ bool SocketConnection::is_connected()
   return !send_ready.failed();
 }
 
+seastar::future<> SocketConnection::send(MessageRef msg)
+{
+  return seastar::with_gate(pending_dispatch, [this, msg=std::move(msg)] {
```

Comment:
why do we need to guard `do_send()` with `pending_dispatch`?

Reply:
@cyx1231st ping?

Reply:
I think it is related to dangling futures.

If the `Connection` is sending messages, e.g. somewhere among `write_message()` related continuations, but suddenly `close()` is called and `Connection` destroyed. There could be segment fault.

When closing a connection, `auto cleanup = [conn = SocketConnectionRef(this)] {};` and proper gating can delay the destructor until all its continuations are finished.